languages/csharp/exercises/concept/attack-of-the-trolls/AttackOfTheTrollsTests.cs

```diff
@@ -3,6 +3,30 @@
 public class PermissionsTests
 {
     [Fact]
+    public void Permission_none_is_an_enum_field()
+    {
+        Assert.Equal("None", Permission.None.ToString());
+    }
+
+    [Fact(Skip = "Remove this Skip property to run this test")]
+    public void Permission_all_is_an_enum_field()
+    {
+        Assert.Equal("All", Permission.All.ToString());
+    }
+
+    [Fact(Skip = "Remove this Skip property to run this test")]
+    public void Permission_is_an_enum_flags_with_read_write()
+    {
+        Assert.Equal("Read, Write", (Permission.Read | Permission.Write).ToString());
+    }
+
+    [Fact(Skip = "Remove this Skip property to run this test")]
+    public void Permission_is_an_enum_flags_with_all()
+    {
+        Assert.Equal("All", (Permission.Read | Permission.Write | Permission.Delete).ToString());
+    }
+
+    [Fact(Skip = "Remove this Skip property to run this test")]
```

Comment:
I'm not sure what these tests add to the existing tests. Tests like `Assert.False(Permissions.Check(Permission.Read | Permission.Write, Permission.Read | Permission.Delete));` already verify the flags aspect of the enum, don't they?

Reply:
This is the premise of the change requested in #2813 .

Those tests show the effect the [Flags] attribute in a test using `ToString()`.

As a side effect, it verifies the enum is properly set up early on, which on my opinion is a good thing, because if the enum is not properly written early on, working with bitwise operators can be scary and draining for a student.

Let me know what you think.

Reply:
Right. I can sort of see the benefit of doing that. I'm not entirely sure that I as a student would know what is actually being tested. Although we usually strive not to use reflection, in this case it _might_ make sense to have this as the first test:

```csharp
Assert.NotEmpty(typeof(Permissions).GetCustomAttributes(typeof(FlagsAttribute)));
```

Reply:
The tests for the `None` and `All` fields seem like they don't offer much value. If the student didn't declare those fields, the tests would not even compile.

```
[Fact]
    public void Permission_none_is_an_enum_field()
    {
        Assert.Equal("None", Permission.None.ToString());
    }

    [Fact(Skip = "Remove this Skip property to run this test")]
    public void Permission_all_is_an_enum_field()
    {
        Assert.Equal("All", Permission.All.ToString());
    }
```

Reply:
Speaking of which, I think the instructions should probably be more clearer on which enum values the student should declare (with their exact names).

Reply:
Custom attribute is in another exercise, so maybe using `GetCustomAttributes` is couple of step ahead of the current exercise.

Tests would not compile even if `Permission_none_is_an_enum_field` is deleted because `Permission.None` is used starting line 50 anyway, so I'm not sure what you mean here.

I added a clear expectation for the enum members in the instruction file.

Reply:
> Tests would not compile even if Permission_none_is_an_enum_field is deleted because Permission.None is used starting line 50 anyway, so I'm not sure what you mean here.

I don't know what this test gives the user. The tests are only run when the code actually compiles. As `Permission.None` is used elsewhere in the file, the code will never compile anyway unless the user has specified the `Permission.None` enum member. So the test won't actually ever fail. That said, it might be useful for CLI-based students to have the tests be a bit explicit. If so, I would then have tests for all the enum members.